lib/adapters/tool/playwright/index.ts

```diff
@@ -0,0 +1,311 @@
+        const startIndex = stdoutByLine.findIndex(v => v === 'Listing tests:');
+        const endIndex = stdoutByLine.findIndex(v => /total: \d+ tests? in \d+ file/i.test(v));
+
+        const tests = stdoutByLine.slice(startIndex + 1, endIndex).map(line => {
+            const [browserName, file, ...titlePath] = line.split(PWT_TITLE_DELIMITER);
+
+            return {
+                browserName: browserName.slice(1, -1),
+                file: file.split(':')[0],
+                title: titlePath.join(DEFAULT_TITLE_DELIMITER),
+                titlePath
+            } as PwtRawTest;
+        });
+
+        return PlaywrightTestCollectionAdapter.create(tests);
+    }
+
+    async run(_testCollection: PlaywrightTestCollectionAdapter, tests: TestSpec[] = []): Promise<boolean> {
+        return this._runTests(tests);
+    }
+
+    async runWithoutRetries(_testCollection: PlaywrightTestCollectionAdapter, tests: TestSpec[] = []): Promise<boolean> {
+        return this._runTests(tests, ['--retries', '0']);
+    }
+
+    private async _runTests(tests: TestSpec[] = [], runArgs: string[] = []): Promise<boolean> {
+        if (!this._reportBuilder || !this._eventSource) {
+            throw new Error('"reportBuilder" and "eventSource" instances must be initialize before run tests');
+        }
+
+        const queue = new PQueue({concurrency: os.cpus().length});
+
+        return new Promise((resolve, reject) => {
+            const args = ([] as string[]).concat(prepareRunArgs(tests, this._configPath, this._haveProjectsInConfig), runArgs);
+            const child = spawn(this._pwtBinaryPath, args, {
+                stdio: ['inherit', 'inherit', 'inherit', 'ipc']
+            });
+
+            ipc.on<PwtEventMessage>(ClientEvents.BEGIN_STATE, (data) => {
+                data.result.status = TestStatus.RUNNING;
+
+                queue.add(async () => {
+                    const testBranch = await registerTestResult(data, this._reportBuilder);
+
+                    this._eventSource.emit(data.event, testBranch);
+                }).catch(reject);
+            }, child);
+
+            ipc.on<PwtEventMessage>(ClientEvents.TEST_RESULT, (data) => {
+                queue.add(async () => {
+                    const testBranch = await registerTestResult(data, this._reportBuilder);
+
+                    this._eventSource.emit(data.event, testBranch);
+                }).catch(reject);
+            }, child);
+
+            ipc.on(ClientEvents.END, (data) => {
+                queue.onIdle().then(() => this._eventSource.emit(data.event));
+            }, child);
+
+            child.on('error', (data) => reject(data));
+
+            child.on('exit', (code) => {
+                queue.onIdle().then(() => resolve(!code));
+            });
+        });
+    }
+
+    // Can't handle test results here because pwt does not provide api for this, so save instances and use them in custom reporter
```

Comment:
By saying "in custom reporter" do you mean those ipc handlers above? (just making sure I got it right)